components/802.15.4_RF/stm-s2lp-rf-driver/source/NanostackRfPhys2lp.cpp

```diff
@@ -374,7 +374,7 @@ static void rf_enable_gpio_interrupt(void)
 
 static void rf_send_command(s2lp_commands_e command)
 {
-    const uint8_t tx[2] = {SPI_CMD, command};
+    const uint8_t tx[2] = {SPI_CMD, (uint8_t)command};
```

Comment:
Is this actually a C++11 issue? As far as I'm aware, as long as all values of the `enum` fit into the destination type, it's not viewed as a narrowing conversion for the initializer-list rules (8.5.4 - "from an integer type or unscoped enumeration type to an integer type that cannot represent all the values of the original type")

(But I do expect most compilers to utter a warning, in any mode, assigning any enum to an int).

Reply:
Oh, are the "values" of the enum for that clause the actual values listed in it, or the possible values of its underlying type?

Going through the rest of the standard, looking for references to "values" of enumerations, I think it's the latter.

Can't find anyone conclusively saying one way or the other out on the net.

Reply:
Having reread C++11 7.2 several times, I can now buy the compiler's interpretation. Just need to be capable of distinguishing "enumerator values" (what you listed) from "values of the enumeration" (values of the underlying type). :) The standard uses both terms in 7.2...

In this case, it seems reasonable to interpret "values of the original type" as the latter, and those are basically implementation-defined, so you can't portably just stick `enum`s into sub-`int` integer initializers any more.

Now, I do hate casts, and the problem is better avoided in C++11 by changing the enumeration to actually specify it's supposed to be `uint8_t`-based. 

Can I suggest instead something like

#if defined __cplusplus && __cplusplus >= 201103
typedef enum : uint8_t {
#else
typedef enum {
#endif
    ...
} s2_lp_commands;

I do also hate `#ifdef`s, but that can be stripped back down when we take C++11 as baseline.

Reply:
This can cause ODR violations if the user decides to define a struct with a field of type `s2_lp_commands` and use it from both C and C++ code, say:
```
struct foo {
    s2_lp_commands cmd;
    char bar;
};
```
Now in C code `bar` has offset 4 in `struct foo`, whereas in C++ code it has offset 1.

Reply:
True (although you'd probably get away with it after #8803).

You'd certainly have to be cautious about doing that sort of change in general.

In this case though it is an internal header used only from the C++ code, so C compatibility isn't an issue. (It's internalness isn't explicit, except for the Mbed OS 3-style code structure where it's living in `source` rather than `stm-s2lp-rf-driver`). I was anticipating the `#ifdef` ultimately dropping so it was just always `enum : uint8_t`, and not C compatible.

Don't know if we need to make the "this is C++" explicit, but we can certainly drop the spurious existing `extern "C"` and my own's `if defined __cplusplus` that suggest it might be C.

Reply:
(Although this is a good example of how you might manage to create a C++03 versus C++11 library incompatibility. I don't think that's a concern in this limited context, but would certainly discourage me from doing similar anywhere more public.)